### PR TITLE
Mongocrypt 324

### PR DESCRIPTION
**Pull request by rozza**
This PR covers a few commits / tickets.

1. 54ba54d MONGOCRYPT-324
    *  Update to use the virtual host style url: https://mciuploads.s3.amazonaws.com/ instead of the deprecated sub directory style
    *  Use the libmongocrypt-all.tar.gz instead of individual zips. Simplifies gradle - no more dynamically having to register multiple tasks per zip to download and process.

    The url it uses reverts the previous change in 6a44fef. I don't think its needed it should still be able to use the githash and have access for publishing. See the [published artifacts here](https://evergreen.mongodb.com/build/libmongocrypt_release_publish_snapshot_3881de79ddcd95939f36100e162e51c3fe7f31ca_22_04_18_20_59_01).  I did have to switch round the evergreen tasks for "publish-snapshot" task relies on the upload-all first, then does the publishing.
2. 310503b update gradle to reduce any tech debt in the build.
3. ce1351d Updated the dependencies
4. 21af59a Use Junit 5 and fix test assertion in `testDataKeyCreation`
5. 399ef8c MONGOCRYPT-324 Test against packaged libraries. Now uses a custom `libmongocrypt-java.tar.gz` file tests using the artefacts in that zip. 
    * In evergreen: 
      * Added `upload-java` task with libmongocrypt-java.tar.gz - contains only the java packages
      * Added `test-java` just runs the tests
      * test.sh - calls `downloadJnaLibs` then runs the tests.
    * In gradle: 
      * testing automatically sets the `jna.debug_load` flag and the `jna.library.path` - defaults to `./build/jnaLibs/<ARCH>/`
6. c237c45 Updated the README.md

Testing: https://spruce.mongodb.com/version/627507d732f4177d537beb45

Failures: 
  * macOS m1 (Apple LLVM) - missing Java? Is it installed?
  * RHEL 7.2 on zSeries - cannot find the artefact.
  * RHEL 7.1 ppc64el - No ppc64el library included - should this be added?
  * Windows 2016 - java.lang.UnsatisfiedLinkError: Error looking up function 'mongocrypt_ctx_provide_kms_providers': The specified procedure could not be found. Any ideas?

### CODE REVIEW

**Comment by rozza** [comment]
I have a few failures - that I can look into, most look simple. Fixing windows I don't fully understand the cause.

Any feedback would be great.

Further improvements would be to automate publishing so that on tagging publishing occurs  and is pushed to sonatype. Same as in the Java driver.

**Review by jyemin — APPROVED** [review]
Love it.

LGTM

**Comment by markbenvenuto** [comment]
LGTM

I cannot use my local build though. It seems it always wants to download.
```$ ./gradlew publishSnapshots -DjnaLibsPath=../../../cmake-build-nocrypto
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :downloadJava FAILED
Download https://mciuploads.s3.amazonaws.com/libmongocrypt/java/f1df0d4aeee56c7f004109862f308ae9b8282b37/libmongocrypt-java.tar.gz

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':downloadJava'.
> A failure occurred while executing de.undercouch.gradle.tasks.download.internal.DefaultWorkerExecutorHelper$DefaultWorkAction
   > de.undercouch.gradle.tasks.download.org.apache.hc.client5.http.ClientProtocolException: Forbidden (HTTP status code: 403, URL: https://mciuploads.s3.amazonaws.com/libmongocrypt/java/f1df0d4aeee56c7f004109862f308ae9b8282b37/libmongocrypt-java.tar.gz)
```

**Review by markbenvenuto — APPROVED** [review]

**Comment by rozza** [comment]
@markbenvenuto 

> I cannot use my local build though. It seems it always wants to download.

Interesting what task were you running?

This is how I tested against a local build:
```
/gradlew check -DjnaLibsPath=/home/rozza/Code/mongodb/libmongocrypt/cmake-build
```

**Comment by markbenvenuto** [comment]
@rozza I cannot repro it today. I do not know why. It builds cleanly but it does not include the `.so` in `jar` file.

But the local build does not include the shared lib in the .jar. I only see the MANIFEST and the .class files in the jar

```$./gradlew publishToMavenLocal  -DjnaLibsPath=/home/mark/src/libmongocrypt/cmake-build-nocrypto
$ unzip -l ./build/libs/mongocrypt-1.5.0-SNAPSHOT.jar | grep -i so
```

File: .evergreen/config.yml [hunk 1]
```diff
@@ -424,28 +422,53 @@ tasks:
...
+    - func: "download tarball"
+      vars: { variant_name: "rhel-67-s390x" }
+    - func: "download tarball"
+      vars: { variant_name: "ubuntu1604-arm64" }
+    - func: "download tarball"
+      vars: { variant_name: "windows-test" }
+    - func: "download tarball"
+      vars: { variant_name: "macos" }
+    - func: "download tarball"
+      vars: { variant_name: "macos_x86_64" }
```
Comment (kevinAlbs) [inline .evergreen/config.yml:457]: > RHEL 7.2 on zSeries - cannot find the artefact.
> RHEL 7.1 ppc64el - No ppc64el library included - should this be added?

Is this resolved by including `rhel72-zseries-test` and `rhel-71-ppc64el` be included in the libmongocrypt-java.tar.gz and adding to the `jnaMapping` in build.gradle.kts?

**Review by kevinAlbs — APPROVED** [review]
Looks good with the addition of the `upload-all` task.

> macOS m1 (Apple LLVM) - missing Java? Is it installed?

I have remote access to a `macos-1100-arm64` host. I see an installation of `/opt/homebrew/Cellar/openjdk/15.0.2`. But I am not sure if it is usable, or if I am using it incorrectly:

```sh
$ JAVA_HOME=/opt/homebrew/Cellar/openjdk/15.0.2 ./gradlew clean check --info -Djna.debug_load=true -Djna.library.path=/Users/dev-user/kevinAlbs/libmongocrypt/cmake-build
Initialized native services in: /Users/dev-user/.gradle/native
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (sharedRuntime.cpp:531), pid=95955, tid=7939
#  Error: ShouldNotReachHere()
#
# JRE version: OpenJDK Runtime Environment (16.0+14) (build 16+14)
# Java VM: OpenJDK 64-Bit Server VM (16+14, mixed mode, tiered, compressed oops, g1 gc, bsd-aarch64)
# Core dump will be written. Default location: /cores/core.95955
#
# An error report file with more information is saved as:
# /Users/dev-user/kevinAlbs/libmongocrypt/bindings/java/mongocrypt/hs_err_pid95955.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
Abort trap: 6
```

> Windows 2016 - java.lang.UnsatisfiedLinkError: Error looking up function 'mongocrypt_ctx_provide_kms_providers': The specified procedure could not be found. Any ideas?

I think that is caused by MONGOCRYPT-428.

File: .evergreen/config.yml [hunk 2]
```diff
@@ -1072,17 +1098,17 @@ buildvariants:
...
   - name: publish-packages
     distros:
     - ubuntu2004-small
 - name: publish-snapshot
   display_name: "Publish"
   run_on: ubuntu1804-test
   tasks:
+    - name: "upload-java"
     - name: "publish-snapshot"
-    - name: "upload-all"
```
[the remark this answers is not in the document]
Comment (rozza) [reply .evergreen/config.yml:1085]: Done.

**Pull request closed without merging**